usr/src/uts/armv8/os/gic.c

```diff
@@ -1,550 +1,139 @@
- * local, and the registers for same are banked per-CPU in the distributor.
- *
- * XXXARM: It's possible these were intended to keep these in sync across all
- * CPUs, but they don't outside of initial config.
- */
-static volatile uint32_t ipriorityr_private[8];
-static volatile uint32_t ienable_private;
-
-/*
- * This is used to record whether an IRQ is edge or level triggered.
- * A set bit in this mask is an edge triggered interrupt, a 0 is level triggered.
- *
- * We use this in `gic_mask_level_irq` and `gic_unmask_level_irq` to decide
- * whether to enable or disable that IRQ, these functions ignore
- * edge-triggered interrupts.  This is set when `gic_config_irq` and never
- * altered (which makes sense).
- */
-static uint32_t intr_cfg[1024 / 32];
-
-/*
- * Protect access to global GIC state.
- * In the current implementation, the distributor.
- */
-lock_t gic_lock;
+static void stub_not_config(void);
+static void stub_setlvlx(int ipl, int irq);
 
-/* Base addresses of the cpu interface and distributor */
-volatile struct gic_cpuif *gic_cpuif;
-volatile struct gic_dist *gic_dist;
+char *gic_module_name = NULL;
 
 /*
- * Mapping from cpuid to GIC target identifier
- * XXXARM: Another place we cap at 8 CPUs
+ * This is terrible, The interrupt trap code needs the GIC
+ * CPU interface base address for GICv2. In implementations
+ * that implement the system register interface we have to
+ * leave this NULL so that the code in ml/exceptions.S can
+ * identify the right thing to do.
  */
-static uint8_t gic_target[8];
+volatile struct gic_cpuif *gic_cpuif = NULL;
 
-/*
- * CPUs for which we have initialized the GIC.  Used to limit IPIs to only
- * those CPUs we can target.
- */
-static cpuset_t gic_cpuset;
+gic_ops_t gic_ops = {
+	.mask_level_irq		= (mask_level_irq_t)stub_not_config,
+	.unmask_level_irq	= (mask_level_irq_t)stub_not_config,
+	.send_ipi		= (send_ipi_t)stub_not_config,
+	.init_primary		= (init_primary_t)stub_not_config,
+	.init_secondary		= (init_secondary_t)stub_not_config,
+	.config_irq		= (config_irq_t)stub_not_config,
+	.num_cpus		= (num_cpus_t)stub_not_config,
+	.addspl			= (addspl_t)stub_not_config,
+	.delspl			= (delspl_t)stub_not_config,
+	.setlvl			= (setlvl_t)stub_not_config,
+	.setlvlx		= stub_setlvlx
+};
 
-/*
- * Enable IRQ in the distributor, which will now be forwarded to a cpu.
- *
- * It is IMPLEMENTATION DEFINED whether SGIs can be enabled or disabled here,
- * we never try.
- */
 static void
-gic_enable_irq(int irq)
+stub_not_config(void)
 {
-	if (irq >= GIC_INTID_MIN_PPI) {
-		gic_dist->isenabler[GICD_IRQ_TO_ISENABLER(irq)] =
-		    GICD_IRQ_TO_ISENABLER_FIELD(irq);
-	}
+	prom_panic("GIC not configured\n");
 }
 
-/*
- * Disable IRQ in the distributor, which will now cease being forwarded to a
- * cpu.
- *
- * It is IMPLEMENTATION DEFINED whether SGIs can be enabled or disabled here,
- * we never try.
- */
 static void
-gic_disable_irq(int irq)
+stub_setlvlx(int ipl, int irq)
```

Comment:
What calls into this before we get the new module loaded?  Do we need it in modstubs?
Do we need the unix modules to depend on the correct GIC implementation instead?
(or a platmod, looking toward the future)?

Reply:
I think a platmod is the right thing to do in the longer term.

I'm not 100% sure what makes all the calls, but based on some tracing I put in there are about 12 calls, all with IPL 15 and IRQ -1, which is effectively what the CPU init code does anyway.

I don't fully understand how modstubs works, but if it simply provides dummy implementations until we can load a module, then it feels better than the manual stubbing I'm doing here.

Reply:
modstubs provide what is in effect an autoload mechanism, we stub out a symbol and record the module it is really in, and when called we try to load the module (or return a default value if we can't).  

I'm not sure it would work here, but it seemed worth wondering.

At present we can't platmod fully because we need serial so very early, and so it has to be baked into unix.

What we could do is make each unix depend upon the correct GIC implementation for that board.

What would be better -- I think -- is to get the GIC available to use out of the feature registers, unfortunately it's a bit spread around and I only plumbed the ones on the PE through into `..._has_feature`.  No need to do that now, while there's only one though.

Reply:
My underlying goal here is to be able to load a GICv3 module where needed, both for FDT and ACPI systems.

I'm working (very slowly, work has been crazy for the last 18 months and is just getting busier) on a EFI/ACPI port - that sort of generic machine would need to decide which implementation to load at runtime. My (known) blocker at the moment is the GICv3 support.

Reply:
Yeah, that's what checking the feature registers would get you.

They're all in `arm_features.h` but they're not all plumbed together, see `cpuid.c`.
You can detect GIC version from the `GICD_PIDR2` register, and then set the features (you'd need to be careful to avoid triggering the CPU mismatch check though).

I'm not saying this is necessary _now_ but it seems the right way to do it programatically (there's another way, once we have ACPI, but not yet...)

Reply:
I've moved GIC detection to do FDT probing from within the GIC wrapper to clean up detection (and make this work a little more like ACPI would).